### ISSUE TITLE
Combination with SimpleFixture

### ISSUE THREAD

**Issue by SeriousM** [ipjohnson/Grace#162]
Hi!
I'm trying to combine Grace with SimpleFixture.
My attempt is as follows:

```cs
public class FixtureBase : SimpleFixture.Fixture
{
  [TestInitialize]
  public void TestSetup()
  {
    var container = new DependencyInjectionContainer();
    container.Configure(reg =>
      reg.ExportFactory<StaticInjectionContext, ILogger>(context =>
        LoggerFactory.GetLogger(context.TargetInfo.InjectionType)));

    Add(new GraceInjectorConvention(container));
    Populate(this);
  }
}

public class GraceInjectorConvention : IConvention
{
  private readonly ILocatorService locatorService;
  public GraceInjectorConvention(ILocatorService locatorService)
  {
    this.locatorService = locatorService;
  }

  public object GenerateData(DataRequest request)
  {
    if (locatorService.TryLocate(request.RequestedType, out object instance))
      return instance;
    return Convention.NoValue;
  }

  // this convention should be ask as last instance. this way the fixture can override types.
  public ConventionPriority Priority { get; } = ConventionPriority.Last;
  public event EventHandler<PriorityChangedEventArgs> PriorityChanged;
}
```
Sorry for the code but it's the smallest example I could create.

The problem is now that `context.TargetInfo.InjectionType` is empty. I guess it's because I use `Populate` from SimpleFixture instead of `Locate` from the container instance.

Is there a way to tell Grace which instance is about to get populated?
I don't need SimpleFixture yet I would have the same issue if I would populate my Fixture properties myself.

Thank you @ipjohnson for Grace, it's awesome :)

/ref: #141

**Comment by ipjohnson** [comment]
@SeriousM I took a look last night and ultimately I think the reason `context.TargetInfo.InjectionType` is empty is because it's only present when the type is being injected into something (that the container knows about) and there is no way to assign it as StaticContext is created when the method is first executed. 

I think what you could do instead is pass the data in through the extra data parameter in locate

```C#
  public object GenerateData(DataRequest request)
  {
    // get injected type from request.ExtraData (is a PropertyInfo or MemberInfo)
    if (locatorService.TryLocate(request.RequestedType, out object instance, new { InjectedType = injectedType }))
      return instance;
    return Convention.NoValue;
  }
```
ILogger registration
```C#
    container.Configure(reg =>
      reg.ExportFactory<StaticInjectionContext, IInjectionContext, ILogger>((context, injectionContext) =>
        LoggerFactory.GetLogger(context.TargetInfo.InjectionType ?? injectionContext.GetExtraData("InjectedType"))));
```

I'll be honest it's a bit hard to follow what you are trying to do with out a working sample.

**Comment by SeriousM** [comment]
> I'll be honest it's a bit hard to follow what you are trying to do with out a working sample.

But you guessed pretty well!
The logger of Serilog (and NLog, log4net, etc.) can be created for a context in which the logger logs. Usually this is the class name so that one is able to determine the source of the log.
To achieve this I create the `ILogger` on the fly in context with the target type.

Your suggestion was a good starting point, but now I have two problems:
* The export factory gets called only once but I need it for every single request
* The extra data is null once I apply a lifestyle. It seems that the execution path somewhere splits into another direction.

I hope you can help me again?

**Comment by ipjohnson** [comment]
@SeriousM you are correct the injection context is not shared through the Singleton lifestyle. That said the logic for the is in the lifestyle itself so it can be overridden. I'll see if I can put together lifestyle that does what you want to do.

**Comment by SeriousM** [comment]
Great, thank you!

**Comment by ipjohnson** [comment]
I've added a test showing custom singleton [here](https://github.com/ipjohnson/Grace/blob/dev/tests/Grace.Tests/DependencyInjection/Lifestyle/CustomSingletonLifestyleTests.cs) that passes the injection context through during construction
